src/binary_search_tree.py

```diff
@@ -0,0 +1,165 @@
+"""Implementation of a binary search tree."""
+
+
+class BinarySearchTree(object):
+    """Binary Search Tree."""
+
+    def __init__(self):
+        """Init bst."""
+        self._values = {}
+        self._root_node = None
+        self._left_depth = 0
+        self._right_depth = 0
+        self._size = 0
+
+    def insert(self, value):
+        """Insert a values, ignores if value is already in bst."""
+        try:
+            if self._values[value]:
+                return value
+        except KeyError:
+            if self._root_node is None:
+                self._root_node = Node(value)
+                self._values.setdefault(value, Node(value))
+                self._size += 1
+                return
+            node = self._root_node
+            depth = 1
+            self._size += 1
+            while True:
+                if value > node.value:
+                    if node.right is None:
+                        node.right = Node(value, depth)
+                        self._values.setdefault(value, node.right)
+                        self._side_depth(value, depth)
+                        return
+                    node = node.right
+                    depth += 1
+                elif value < node.value:
+                    if node.left is None:
+                        node.left = Node(value, depth)
+                        self._values.setdefault(value, node.left)
+                        self._side_depth(value, depth)
+                        return
+                    node = node.left
+                    depth += 1
+
+    def _side_depth(self, value, depth):
+        """Helper function to find what side depth to increment."""
+        if value > self._root_node.value:
+            side = "right"
+        else:
+            side = "left"
+        if side == "left" and depth > self._left_depth:
```

Comment:
I think it might be beneficial in the future if you look into using getattr and setattr